crates/deltalake-core/src/operations/constraints.rs

```diff
@@ -0,0 +1,263 @@
+            state: None,
+        }
+    }
+
+    /// Specify the constraint to be added
+    pub fn with_constraint<S: Into<String>, E: Into<Expression>>(
+        mut self,
+        column: S,
+        expression: E,
+    ) -> Self {
+        self.name = Some(column.into());
+        self.expr = Some(expression.into());
+        self
+    }
+
+    /// Specify the datafusion session context
+    pub fn with_session_state(mut self, state: SessionState) -> Self {
+        self.state = Some(state);
+        self
+    }
+}
+
+impl std::future::IntoFuture for ConstraintBuilder {
+    type Output = DeltaResult<DeltaTable>;
+
+    type IntoFuture = BoxFuture<'static, Self::Output>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        let this = self;
+
+        Box::pin(async move {
+            if this.name.is_none() {
+                return Err(DeltaTableError::Generic("No name provided".to_string()));
+            } else if this.expr.is_none() {
+                return Err(DeltaTableError::Generic(
+                    "No expression provided".to_string(),
+                ));
+            }
+            let name = this.name.unwrap();
+            let expr = match this.expr {
+                Some(Expression::String(s)) => s,
+                Some(Expression::DataFusion(e)) => e.to_string(),
+                None => unreachable!(),
+            };
+
+            let mut metadata = this
+                .snapshot
+                .metadata()
+                .ok_or(DeltaTableError::NoMetadata)?
+                .clone();
+            let configuration_key = format!("delta.constraints.{}", name);
+
+            if metadata.configuration.contains_key(&configuration_key) {
+                return Err(DeltaTableError::Generic(format!(
+                    "Constraint with name: {} already exists, expr: {}",
+                    name, expr
+                )));
+            }
+
+            let state = this.state.unwrap_or_else(|| {
+                let session = SessionContext::new();
+                register_store(this.log_store.clone(), session.runtime_env());
+                session.state()
+            });
+
+            // Checker built here with the one time constraint to check.
+            let checker = DeltaDataChecker::with_constraints(vec![Constraint::new("*", &expr)]);
+
+            let files_to_check =
+                find_files(&this.snapshot, this.log_store.clone(), &state, None).await?;
+            let scan = DeltaScanBuilder::new(&this.snapshot, this.log_store.clone(), &state)
+                .with_files(&files_to_check.candidates)
+                .build()
+                .await?;
+
+            let task_ctx = Arc::new(TaskContext::from(&state));
+            let record_stream: SendableRecordBatchStream = scan.execute(0, task_ctx)?;
+            let records = collect_sendable_stream(record_stream).await?;
+
+            for batch in records {
+                checker.check_batch(&batch).await?;
+            }
```

Comment:
Should look similar to what we do when we write the stream to disk - sans the writing stuff :)

https://github.com/delta-io/delta-rs/blob/e54996ae3bc4a0d1b838e6a584b2a636171f387e/crates/deltalake-core/src/operations/write.rs#L324-L351

To avoid having to iterate through all the partitions in a loop, You could also wrap it in a `CoalesceExec`, where you would always end up with a single partition.

Reply:
I used exactly this to reflect the ask. :-)